File: app/src/main/java/eu/kanade/tachiyomi/data/source/online/english/Batoto.java
package eu.kanade.tachiyomi.data.source.online.english;

import android.content.Context;
import android.net.Uri;
import android.text.TextUtils;

import com.squareup.okhttp.FormEncodingBuilder;
import com.squareup.okhttp.Headers;
import com.squareup.okhttp.Response;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.net.HttpCookie;
import java.net.URI;
import java.net.URISyntaxException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import eu.kanade.tachiyomi.data.database.models.Chapter;
import eu.kanade.tachiyomi.data.database.models.Manga;
import eu.kanade.tachiyomi.data.source.SourceManager;
import eu.kanade.tachiyomi.data.source.base.LoginSource;
import eu.kanade.tachiyomi.data.source.model.MangasPage;
import eu.kanade.tachiyomi.data.source.model.Page;
import eu.kanade.tachiyomi.util.Parser;
import rx.Observable;

public class Batoto extends LoginSource {

    public static final String NAME = "Batoto (EN)";
    public static final String BASE_URL = "http://bato.to";
    public static final String POPULAR_MANGAS_URL = BASE_URL + "/search_ajax?order_cond=views&order=desc&p=%d";
    public static final String SEARCH_URL = BASE_URL + "/search_ajax?name=%s&p=%s";
    public static final String CHAPTER_URL = "/areader?id=%s&p=1";
    public static final String PAGE_URL = BASE_URL + "/areader?id=%s&p=%s";
    public static final String MANGA_URL = "/comic_pop?id=%s";
    public static final String LOGIN_URL = BASE_URL + "/forums/index.php?app=core&module=global&section=login";

    private Pattern datePattern;
    private Map<String, Integer> dateFields;

    public Batoto(Context context) {
        super(context);

        datePattern = Pattern.compile("(\\d+|A|An)\\s+(.*?)s? ago.*");
        dateFields = new HashMap<String, Integer>() {{
            put("second", Calendar.SECOND);
            put("minute", Calendar.MINUTE);
            put("hour",   Calendar.HOUR);
            put("day",    Calendar.DATE);
            put("week",   Calendar.WEEK_OF_YEAR);
            put("month",  Calendar.MONTH);
            put("year",   Calendar.YEAR);
        }};
    }

    @Override
    public String getName() {
        return NAME;
    }

    @Override
    public int getId() {
        return SourceManager.BATOTO;
    }

    @Override
    public String getBaseUrl() {
        return BASE_URL;
    }

    @Override
    protected Headers.Builder headersBuilder() {
        Headers.Builder builder = super.headersBuilder();
        builder.add("Cookie", "lang_option=English");
        builder.add("Referer", "http://bato.to/reader");
        return builder;
    }

    @Override
    public String getInitialPopularMangasUrl() {
        return String.format(POPULAR_MANGAS_URL, 1);
    }

    @Override
    public String getInitialSearchUrl(String query) {
        return String.format(SEARCH_URL, Uri.encode(query), 1);
    }

    @Override
    protected String overrideMangaUrl(String defaultMangaUrl) {
        String mangaId = defaultMangaUrl.substring(defaultMangaUrl.lastIndexOf("r") + 1);
        return String.format(MANGA_URL, mangaId);
    }

    @Override
    protected String overrideChapterUrl(String defaultPageUrl) {
        String id = defaultPageUrl.substring(defaultPageUrl.indexOf("#") + 1);
        return String.format(CHAPTER_URL, id);
    }

    @Override
    protected String overridePageUrl(String defaultPageUrl) {
        int start = defaultPageUrl.indexOf("#") + 1;
        int end = defaultPageUrl.indexOf("_", start);
        String id = defaultPageUrl.substring(start, end);
        return String.format(PAGE_URL, id, defaultPageUrl.substring(end+1));
    }

    private List<Manga> parseMangasFromHtml(Document parsedHtml) {
        List<Manga> mangaList = new ArrayList<>();

        if (!parsedHtml.text().contains("No (more) comics found!")) {
            for (Element currentHtmlBlock : parsedHtml.select("tr:not([id]):not([class])")) {
                Manga manga = constructMangaFromHtmlBlock(currentHtmlBlock);
                mangaList.add(manga);
            }
        }
        return mangaList;
    }

    @Override
    protected List<Manga> parsePopularMangasFromHtml(Document parsedHtml) {
        return parseMangasFromHtml(parsedHtml);
    }

    @Override
    protected String parseNextPopularMangasUrl(Document parsedHtml, MangasPage page) {
        Element next = Parser.element(parsedHtml, "#show_more_row");
        return next != null ? String.format(POPULAR_MANGAS_URL, page.page + 1) : null;
    }

    @Override
    protected List<Manga> parseSearchFromHtml(Document parsedHtml) {
        return parseMangasFromHtml(parsedHtml);
    }

    private Manga constructMangaFromHtmlBlock(Element htmlBlock) {
        Manga manga = new Manga();
        manga.source = getId();

        Element urlElement = Parser.element(htmlBlock, "a[href^=http://bato.to]");
        if (urlElement != null) {
            manga.setUrl(urlElement.attr("href"));
            manga.title = urlElement.text().trim();
        }
        return manga;
    }

    @Override
    protected String parseNextSearchUrl(Document parsedHtml, MangasPage page, String query) {
        Element next = Parser.element(parsedHtml, "#show_more_row");
        return next != null ? String.format(SEARCH_URL, query, page.page + 1) : null;
    }

    @Override
    protected Manga parseHtmlToManga(String mangaUrl, String unparsedHtml) {
        Document parsedDocument = Jsoup.parse(unparsedHtml);

        Element tbody = parsedDocument.select("tbody").first();
        Element artistElement = tbody.select("tr:contains(Author/Artist:)").first();
        Elements genreElements = tbody.select("tr:contains(Genres:) img");

        Manga manga = Manga.create(mangaUrl);
        manga.author = Parser.text(artistElement, "td:eq(1)");
        manga.artist = Parser.text(artistElement, "td:eq(2)", manga.author);
        manga.description = Parser.text(tbody, "tr:contains(Description:) > td:eq(1)");
        manga.thumbnail_url = Parser.src(parsedDocument, "img[src^=http://img.bato.to/forums/uploads/]");
        manga.status = parseStatus(Parser.text(parsedDocument, "tr:contains(Status:) > td:eq(1)"));

        if (!genreElements.isEmpty()) {
            List<String> genres = new ArrayList<>();
            for (Element element : genreElements) {
                genres.add(element.attr("alt"));
            }
            manga.genre = TextUtils.join(", ", genres);
        }

        manga.initialized = true;
        return manga;
    }

    private int parseStatus(String status) {
        switch (status) {
            case "Ongoing":
                return Manga.ONGOING;
            case "Complete":
                return Manga.COMPLETED;
            default:
                return Manga.UNKNOWN;
        }
    }

    @Override
    protected List<Chapter> parseHtmlToChapters(String unparsedHtml) {
        Document parsedDocument = Jsoup.parse(unparsedHtml);

        List<Chapter> chapterList = new ArrayList<>();

        Elements chapterElements = parsedDocument.select("tr.row.lang_English.chapter_row");
        for (Element chapterElement : chapterElements) {
            Chapter chapter = constructChapterFromHtmlBlock(chapterElement);
            chapterList.add(chapter);
        }
        return chapterList;

    }

    private Chapter constructChapterFromHtmlBlock(Element chapterElement) {
        Chapter chapter = Chapter.create();

        Element urlElement = chapterElement.select("a[href^=http://bato.to/reader").first();
        Element dateElement = chapterElement.select("td").get(4);

        if (urlElement != null) {
            String fieldUrl = urlElement.attr("href");
            chapter.setUrl(fieldUrl);
            chapter.name = urlElement.text().trim();
        }
        if (dateElement != null) {
            chapter.date_upload = parseDateFromElement(dateElement);
        }
        return chapter;
    }

    private long parseDateFromElement(Element dateElement) {
        String dateAsString = dateElement.text();

        Date date;
        try {
            date = new SimpleDateFormat("dd MMMMM yyyy - hh:mm a", Locale.ENGLISH).parse(dateAsString);
        } catch (ParseException e) {
            Matcher m = datePattern.matcher(dateAsString);

            if (m.matches()) {
                String number = m.group(1);
                int amount = number.contains("A") ? 1 : Integer.parseInt(m.group(1));
                String unit = m.group(2);

                Calendar cal = Calendar.getInstance();
                // Not an error
                cal.add(dateFields.get(unit), -amount);
                date = cal.getTime();
            } else {
                return 0;
            }
        }
        return date.getTime();
    }

    @Override
    protected List<String> parseHtmlToPageUrls(String unparsedHtml) {
        Document parsedDocument = Jsoup.parse(unparsedHtml);

        List<String> pageUrlList = new ArrayList<>();

        Element selectElement = Parser.element(parsedDocument, "#page_select");
        if (selectElement != null) {
            for (Element pageUrlElement : selectElement.select("option")) {
                pageUrlList.add(pageUrlElement.attr("value"));
            }
        } else {
            // For webtoons in one page
            for (int i = 0; i < parsedDocument.select("div > img").size(); i++) {
                pageUrlList.add("");
            }
        }

        return pageUrlList;
    }

    @Override
    protected List<Page> parseFirstPage(List<Page> pages, String unparsedHtml) {
        if (!unparsedHtml.contains("Want to see this chapter per page instead?")) {
            String firstImage = parseHtmlToImageUrl(unparsedHtml);
            pages.get(0).setImageUrl(firstImage);
        } else {
            // For webtoons in one page
            Document parsedDocument = Jsoup.parse(unparsedHtml);
            Elements imageUrls = parsedDocument.select("div > img");
            for (int i = 0; i < pages.size(); i++) {
                pages.get(i).setImageUrl(imageUrls.get(i).attr("src"));
            }
        }
        return pages;
    }

    @Override
    protected String parseHtmlToImageUrl(String unparsedHtml) {
        int beginIndex = unparsedHtml.indexOf("<img id=\"comic_page\"");
        int endIndex = unparsedHtml.indexOf("</a>", beginIndex);
        String trimmedHtml = unparsedHtml.substring(beginIndex, endIndex);

        Document parsedDocument = Jsoup.parse(trimmedHtml);
        Element imageElement = parsedDocument.getElementById("comic_page");
        return imageElement.attr("src");
    }

    @Override
    public Observable<Boolean> login(String username, String password) {
        return networkService.getStringResponse(LOGIN_URL, requestHeaders, null)
                .flatMap(response -> doLogin(response, username, password))
                .map(this::isAuthenticationSuccessful);
    }

    private Observable<Response> doLogin(String response, String username, String password) {
        Document doc = Jsoup.parse(response);
        Element form = doc.select("#login").first();
        String postUrl = form.attr("action");

        FormEncodingBuilder formBody = new FormEncodingBuilder();
        Element authKey = form.select("input[name=auth_key]").first();

        formBody.add(authKey.attr("name"), authKey.attr("value"));
        formBody.add("ips_username", username);
        formBody.add("ips_password", password);
        formBody.add("invisible", "1");
        formBody.add("rememberMe", "1");

        return networkService.postData(postUrl, formBody.build(), requestHeaders);
    }

    @Override
    protected boolean isAuthenticationSuccessful(Response response) {
        return response.priorResponse() != null && response.priorResponse().code() == 302;
    }

    @Override
    public boolean isLogged() {
        try {
            for ( HttpCookie cookie : networkService.getCookies().get(new URI(BASE_URL)) ) {
                if (cookie.getName().equals("pass_hash"))
                    return true;
            }

        } catch (URISyntaxException e) {
            e.printStackTrace();
        }
        return false;
    }

    @Override
    public Observable<List<Chapter>> pullChaptersFromNetwork(String mangaUrl) {
        Observable<List<Chapter>> observable;
        if (!isLogged()) {
            observable = login(prefs.getSourceUsername(this), prefs.getSourcePassword(this))
                    .flatMap(result -> super.pullChaptersFromNetwork(mangaUrl));
        }
        else {
            observable = super.pullChaptersFromNetwork(mangaUrl);
        }
        return observable;
    }

}

File: app/src/main/java/eu/kanade/tachiyomi/data/source/online/english/Batoto.kt
package eu.kanade.tachiyomi.data.source.online.english

import android.text.Html
import eu.kanade.tachiyomi.data.database.models.Chapter
import eu.kanade.tachiyomi.data.database.models.Manga
import eu.kanade.tachiyomi.data.network.GET
import eu.kanade.tachiyomi.data.network.POST
import eu.kanade.tachiyomi.data.network.asObservable
import eu.kanade.tachiyomi.data.source.model.MangasPage
import eu.kanade.tachiyomi.data.source.model.Page
import eu.kanade.tachiyomi.data.source.online.LoginSource
import eu.kanade.tachiyomi.data.source.online.ParsedOnlineSource
import eu.kanade.tachiyomi.util.asJsoup
import eu.kanade.tachiyomi.util.selectText
import okhttp3.FormBody
import okhttp3.HttpUrl
import okhttp3.Request
import okhttp3.Response
import org.jsoup.nodes.Document
import org.jsoup.nodes.Element
import rx.Observable
import java.net.URI
import java.text.ParseException
import java.text.SimpleDateFormat
import java.util.*
import java.util.regex.Pattern

class Batoto(override val id: Int) : ParsedOnlineSource(), LoginSource {

    override val name = "Batoto"

    override val baseUrl = "http://bato.to"

    override val lang = "en"

    override val supportsLatest = true

    private val datePattern = Pattern.compile("(\\d+|A|An)\\s+(.*?)s? ago.*")

    private val dateFields = HashMap<String, Int>().apply {
        put("second", Calendar.SECOND)
        put("minute", Calendar.MINUTE)
        put("hour", Calendar.HOUR)
        put("day", Calendar.DATE)
        put("week", Calendar.WEEK_OF_YEAR)
        put("month", Calendar.MONTH)
        put("year", Calendar.YEAR)
    }

    private val staffNotice = Pattern.compile("=+Batoto Staff Notice=+([^=]+)==+", Pattern.CASE_INSENSITIVE)

    override fun headersBuilder() = super.headersBuilder()
            .add("Cookie", "lang_option=English")

    private val pageHeaders = super.headersBuilder()
            .add("Referer", "http://bato.to/reader")
            .build()

    override fun popularMangaInitialUrl() = "$baseUrl/search_ajax?order_cond=views&order=desc&p=1"

    override fun latestUpdatesInitialUrl() = "$baseUrl/search_ajax?order_cond=update&order=desc&p=1"

    override fun popularMangaParse(response: Response, page: MangasPage) {
        val document = response.asJsoup()
        for (element in document.select(popularMangaSelector())) {
            Manga.create(id).apply {
                popularMangaFromElement(element, this)
                page.mangas.add(this)
            }
        }

        page.nextPageUrl = document.select(popularMangaNextPageSelector()).first()?.let {
            "$baseUrl/search_ajax?order_cond=views&order=desc&p=${page.page + 1}"
        }
    }

    override fun latestUpdatesParse(response: Response, page: MangasPage) {
        val document = response.asJsoup()
        for (element in document.select(latestUpdatesSelector())) {
            Manga.create(id).apply {
                latestUpdatesFromElement(element, this)
                page.mangas.add(this)
            }
        }

        page.nextPageUrl = document.select(latestUpdatesNextPageSelector()).first()?.let {
            "$baseUrl/search_ajax?order_cond=update&order=desc&p=${page.page + 1}"
        }
    }

    override fun popularMangaSelector() = "tr:has(a)"

    override fun latestUpdatesSelector() = "tr:has(a)"

    override fun popularMangaFromElement(element: Element, manga: Manga) {
        element.select("a[href^=http://bato.to]").first().let {
            manga.setUrlWithoutDomain(it.attr("href"))
            manga.title = it.text().trim()
        }
    }

    override fun latestUpdatesFromElement(element: Element, manga: Manga) {
        popularMangaFromElement(element, manga)
    }

    override fun popularMangaNextPageSelector() = "#show_more_row"

    override fun latestUpdatesNextPageSelector() = "#show_more_row"

    override fun searchMangaInitialUrl(query: String, filters: List<Filter<*>>) = searchMangaUrl(query, filters, 1)

    private fun searchMangaUrl(query: String, filterStates: List<Filter<*>>, page: Int): String {
        val url = HttpUrl.parse("$baseUrl/search_ajax").newBuilder()
        if (!query.isEmpty()) url.addQueryParameter("name", query).addQueryParameter("name_cond", "c")
        var genres = ""
        for (filter in if (filterStates.isEmpty()) filters else filterStates) {
            when (filter) {
                is Status -> if (filter.state != Filter.TriState.STATE_IGNORE) {
                    url.addQueryParameter("completed", if (filter.state == Filter.TriState.STATE_EXCLUDE) "i" else "c")
                }
                is Genre -> if (filter.state != Filter.TriState.STATE_IGNORE) {
                    genres += (if (filter.state == Filter.TriState.STATE_EXCLUDE) ";e" else ";i") + filter.id
                }
                is TextField -> {
                    if (!filter.state.isEmpty()) url.addQueryParameter(filter.key, filter.state)
                }
                is ListField -> {
                    val sel = filter.values[filter.state].value
                    if (!sel.isEmpty()) url.addQueryParameter(filter.key, sel)
                }
                is Flag -> {
                    val sel = if (filter.state) filter.valTrue else filter.valFalse
                    if (!sel.isEmpty()) url.addQueryParameter(filter.key, sel)
                }
            }
        }
        if (!genres.isEmpty()) url.addQueryParameter("genres", genres)
        url.addQueryParameter("p", page.toString())
        return url.toString()
    }

    override fun searchMangaRequest(page: MangasPage, query: String, filters: List<Filter<*>>): Request {
        if (page.page == 1) {
            page.url = searchMangaInitialUrl(query, filters)
        }
        return GET(page.url, headers)
    }

    override fun searchMangaParse(response: Response, page: MangasPage, query: String, filters: List<Filter<*>>) {
        val document = response.asJsoup()
        for (element in document.select(searchMangaSelector())) {
            Manga.create(id).apply {
                searchMangaFromElement(element, this)
                page.mangas.add(this)
            }
        }

        page.nextPageUrl = document.select(searchMangaNextPageSelector()).first()?.let {
            searchMangaUrl(query, filters, page.page + 1)
        }
    }

    override fun searchMangaSelector() = popularMangaSelector()

    override fun searchMangaFromElement(element: Element, manga: Manga) {
        popularMangaFromElement(element, manga)
    }

    override fun searchMangaNextPageSelector() = popularMangaNextPageSelector()

    override fun mangaDetailsRequest(manga: Manga): Request {
        val mangaId = manga.url.substringAfterLast("r")
        return GET("$baseUrl/comic_pop?id=$mangaId", headers)
    }

    override fun mangaDetailsParse(document: Document, manga: Manga) {
        val tbody = document.select("tbody").first()
        val artistElement = tbody.select("tr:contains(Author/Artist:)").first()

        manga.author = artistElement.selectText("td:eq(1)")
        manga.artist = artistElement.selectText("td:eq(2)") ?: manga.author
        manga.description = tbody.selectText("tr:contains(Description:) > td:eq(1)")
        manga.thumbnail_url = document.select("img[src^=http://img.bato.to/forums/uploads/]").first()?.attr("src")
        manga.status = parseStatus(document.selectText("tr:contains(Status:) > td:eq(1)"))
        manga.genre = tbody.select("tr:contains(Genres:) img").map { it.attr("alt") }.joinToString(", ")
    }

    private fun parseStatus(status: String?) = when (status) {
        "Ongoing" -> Manga.ONGOING
        "Complete" -> Manga.COMPLETED
        else -> Manga.UNKNOWN
    }

    override fun chapterListParse(response: Response, chapters: MutableList<Chapter>) {
        val body = response.body().string()
        val matcher = staffNotice.matcher(body)
        if (matcher.find()) {
            val notice = Html.fromHtml(matcher.group(1)).toString().trim()
            throw Exception(notice)
        }

        val document = response.asJsoup(body)

        for (element in document.select(chapterListSelector())) {
            Chapter.create().apply {
                chapterFromElement(element, this)
                chapters.add(this)
            }
        }
    }

    override fun chapterListSelector() = "tr.row.lang_English.chapter_row"

    override fun chapterFromElement(element: Element, chapter: Chapter) {
        val urlElement = element.select("a[href^=http://bato.to/reader").first()

        chapter.setUrlWithoutDomain(urlElement.attr("href"))
        chapter.name = urlElement.text()
        chapter.date_upload = element.select("td").getOrNull(4)?.let {
            parseDateFromElement(it)
        } ?: 0
    }

    private fun parseDateFromElement(dateElement: Element): Long {
        val dateAsString = dateElement.text()

        var date: Date
        try {
            date = SimpleDateFormat("dd MMMMM yyyy - hh:mm a", Locale.ENGLISH).parse(dateAsString)
        } catch (e: ParseException) {
            val m = datePattern.matcher(dateAsString)

            if (m.matches()) {
                val number = m.group(1)
                val amount = if (number.contains("A")) 1 else Integer.parseInt(m.group(1))
                val unit = m.group(2)

                date = Calendar.getInstance().apply {
                    add(dateFields[unit]!!, -amount)
                }.time
            } else {
                return 0
            }
        }

        return date.time
    }

    override fun pageListRequest(chapter: Chapter): Request {
        val id = chapter.url.substringAfterLast("#")
        return GET("$baseUrl/areader?id=$id&p=1", pageHeaders)
    }

    override fun pageListParse(document: Document, pages: MutableList<Page>) {
        val selectElement = document.select("#page_select").first()
        if (selectElement != null) {
            for ((i, element) in selectElement.select("option").withIndex()) {
                pages.add(Page(i, element.attr("value")))
            }
            pages.getOrNull(0)?.imageUrl = imageUrlParse(document)
        } else {
            // For webtoons in one page
            for ((i, element) in document.select("div > img").withIndex()) {
                pages.add(Page(i, "", element.attr("src")))
            }
        }
    }

    override fun imageUrlRequest(page: Page): Request {
        val pageUrl = page.url
        val start = pageUrl.indexOf("#") + 1
        val end = pageUrl.indexOf("_", start)
        val id = pageUrl.substring(start, end)
        return GET("$baseUrl/areader?id=$id&p=${pageUrl.substring(end + 1)}", pageHeaders)
    }

    override fun imageUrlParse(document: Document): String {
        return document.select("#comic_page").first().attr("src")
    }

    override fun login(username: String, password: String) =
            client.newCall(GET("$baseUrl/forums/index.php?app=core&module=global&section=login", headers))
                    .asObservable()
                    .flatMap { doLogin(it, username, password) }
                    .map { isAuthenticationSuccessful(it) }

    private fun doLogin(response: Response, username: String, password: String): Observable<Response> {
        val doc = response.asJsoup()
        val form = doc.select("#login").first()
        val url = form.attr("action")
        val authKey = form.select("input[name=auth_key]").first()

        val payload = FormBody.Builder().apply {
            add(authKey.attr("name"), authKey.attr("value"))
            add("ips_username", username)
            add("ips_password", password)
            add("invisible", "1")
            add("rememberMe", "1")
        }.build()

        return client.newCall(POST(url, headers, payload)).asObservable()
    }

    override fun isAuthenticationSuccessful(response: Response) =
            response.priorResponse() != null && response.priorResponse().code() == 302

    override fun isLogged(): Boolean {
        return network.cookies.get(URI(baseUrl)).any { it.name() == "pass_hash" }
    }

    override fun fetchChapterList(manga: Manga): Observable<List<Chapter>> {
        if (!isLogged()) {
            val username = preferences.sourceUsername(this)
            val password = preferences.sourcePassword(this)

            if (username.isNullOrEmpty() || password.isNullOrEmpty()) {
                return Observable.error(Exception("User not logged"))
            } else {
                return login(username, password).flatMap { super.fetchChapterList(manga) }
            }

        } else {
            return super.fetchChapterList(manga)
        }
    }

    private data class ListValue(val name: String, val value: String) {
        override fun toString(): String = name
    }

    private class Status() : Filter.TriState("Completed")
    private class Genre(name: String, val id: Int) : Filter.TriState(name)
    private class TextField(name: String, val key: String) : Filter.Text(name)
    private class ListField(name: String, val key: String, values: Array<ListValue>, state: Int = 0) : Filter.List<ListValue>(name, values, state)
    private class Flag(name: String, val key: String, val valTrue: String, val valFalse: String) : Filter.CheckBox(name)

    // [...document.querySelectorAll("#advanced_options div.genre_buttons")].map((el,i) => {
    //     const onClick=el.getAttribute('onclick');const id=onClick.substr(14,onClick.length-16);return `Genre("${el.textContent.trim()}", ${id})`
    // }).join(',\n')
    // on https://bato.to/search
    override fun getFilterList(): List<Filter<*>> = listOf(
            TextField("Author", "artist_name"),
            ListField("Type", "type", arrayOf(ListValue("Any", ""), ListValue("Manga (Jp)", "jp"), ListValue("Manhwa (Kr)", "kr"), ListValue("Manhua (Cn)", "cn"), ListValue("Artbook", "ar"), ListValue("Other", "ot"))),
            Status(),
            Flag("Exclude mature", "mature", "m", ""),
            ListField("Order by", "order_cond", arrayOf(ListValue("Title", "title"), ListValue("Author", "author"), ListValue("Artist", "artist"), ListValue("Rating", "rating"), ListValue("Views", "views"), ListValue("Last Update", "update")), 4),
            Flag("Ascending order", "order", "asc", "desc"),
            Filter.Header("Genres"),
            ListField("Inclusion mode", "genre_cond", arrayOf(ListValue("And (all selected genres)", "and"), ListValue("Or (any selected genres) ", "or"))),
            Genre("4-Koma", 40),
            Genre("Action", 1),
            Genre("Adventure", 2),
            Genre("Award Winning", 39),
            Genre("Comedy", 3),
            Genre("Cooking", 41),
            Genre("Doujinshi", 9),
            Genre("Drama", 10),
            Genre("Ecchi", 12),
            Genre("Fantasy", 13),
            Genre("Gender Bender", 15),
            Genre("Harem", 17),
            Genre("Historical", 20),
            Genre("Horror", 22),
            Genre("Josei", 34),
            Genre("Martial Arts", 27),
            Genre("Mecha", 30),
            Genre("Medical", 42),
            Genre("Music", 37),
            Genre("Mystery", 4),
            Genre("Oneshot", 38),
            Genre("Psychological", 5),
            Genre("Romance", 6),
            Genre("School Life", 7),
            Genre("Sci-fi", 8),
            Genre("Seinen", 32),
            Genre("Shoujo", 35),
            Genre("Shoujo Ai", 16),
            Genre("Shounen", 33),
            Genre("Shounen Ai", 19),
            Genre("Slice of Life", 21),
            Genre("Smut", 23),
            Genre("Sports", 25),
            Genre("Supernatural", 26),
            Genre("Tragedy", 28),
            Genre("Webtoon", 36),
            Genre("Yaoi", 29),
            Genre("Yuri", 31),
            Genre("[no chapters]", 44)
    )

}